### ISSUE TITLE
panic in `merge_thread_0`, v0.19.0 from crates.io

### ISSUE THREAD

**Issue by eeeebbbbrrrr** [quickwit-oss/tantivy-cli#94]
**Describe the bug**
- What did you do?

I have the "pile" dataset locally.  It's about 1.5T (210M docs) of jsonl data.  The input data has only 1 property named `"text"`.  I created a new tantivy index:

```shell
$ cargo install tantivy-cli
$ tantivy new --index tantivy-pile
```

And answered the questions as so:

```
Creating new index 
First define its schema! 



New field name  ? text
Choose Field Type (Text/u64/i64/f64/Date/Facet/Bytes/Json/bool/IpAddr) ? Text
Should the field be stored (Y/N) ? y
Should the field be fast (Y/N) ? y
Should the field be indexed (Y/N) ? y
Should the term be tokenized? (Y/N) ? y
Should the term frequencies (per doc) be in the index (Y/N) ? y
Should the term positions (per doc) be in the index (Y/N) ? y
Add another field (Y/N) ? n

[
  {
    "name": "text",
    "type": "text",
    "options": {
      "indexing": {
        "record": "position",
        "fieldnorms": true,
        "tokenizer": "en_stem"
      },
      "stored": true,
      "fast": true
    }
  }
]
```

- What happened?

Next, I ran `cat ./pile-full.jsonl | tantivy index --index tantivy-pile` and waited maybe 10 minutes?  Then this:

```
...
42136080 docs / hour 78.23 Mb/s
4881534 Docs
37922192 docs / hour 69.05 Mb/s
4891111 Docs
34472304 docs / hour 66.04 Mb/s
4901931 Docs
38951880 docs / hour 75.06 Mb/s
4910233 Docs
29886992 docs / hour 51.27 Mb/s
4921717 Docs
41341904 docs / hour 66.33 Mb/s
4931622 Docs
35656184 docs / hour 62.02 Mb/s
4943683 Docs
43417688 docs / hour 81.10 Mb/s
4952702 Docs
32462428 docs / hour 55.23 Mb/s
4963757 Docs
39797048 docs / hour 70.79 Mb/s
4974458 Docs
38522676 docs / hour 68.65 Mb/s
4984414 Docs
35840344 docs / hour 65.31 Mb/s
4995110 Docs
38505520 docs / hour 70.86 Mb/s
thread 'merge_thread_0' panicked at 'index out of bounds: the len is 3860946 but the index is 3985188', /home/zombodb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tantivy-0.19.2/src/indexer/merger.rs:633:40
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
5003668 Docs
30807754 docs / hour 50.13 Mb/s
Rayon: detected unexpected panic; aborting
[1]    771398 broken pipe                    cat pile-full.jsonl | 
       771399 IOT instruction (core dumped)  tantivy index -i tantivy-pile
```

- What was expected?

Based on the math, I expected it to finish in about 5.25 hours.

**Which version of tantivy are you using?**
If "master",  ideally give the specific sha1 revision.

```shell
$ tantivy -V
Tantivy 0.19.0
```

**To Reproduce**

I didn't dig into the code, but I'd have to guess it might be data-related.  I'd be happy to work with someone offline to send either my full dataset or the first 6-10M rows.  Let me know if that'd be helpful.

**Comment by eeeebbbbrrrr** [comment]
Tried again after compiling from this repo (5666983d758c4061070286a0bca55f6e9bff73d8) and it's progressed past where it previously crashed.  

I'll report back this evening if it successfully finished or not.

**Comment by eeeebbbbrrrr** [comment]
Shortly after that comment above, the linux OOM killer got the process, but I thought *maaaybe* that could have been due to me still using the computer while it was running.  This machine is an AMD Ryzen 9 7950X3D 16-Core Processor with 126G of ram -- it's pretty beefy.

So I kicked it off again right before I went to bed and while the original panic I reported seems to be gone, the linux OOM killer still got it after ~90 minutes.  This time the computer was otherwise completely idle.

```
$ tantivy -V
Tantivy 0.20.0
$ date; cat pile-full.jsonl | tantivy index -i tantivy-pile; date
Tue Aug 29 00:37:34 EDT 2023
22144 Docs
78889584 docs / hour 136.94 Mb/s
35584 Docs
48203000 docs / hour 82.02 Mb/s
48384 Docs
...
610018 docs / hour 0.82 Mb/s
44332032 Docs
159151 docs / hour 0.26 Mb/s
44332160 Docs
95349 docs / hour 0.21 Mb/s
44332288 Docs
100721 docs / hour 0.15 Mb/s
[1]    852337 broken pipe  cat pile-full.jsonl | 
       852338 killed       tantivy index -i tantivy-pile
Tue Aug 29 01:54:44 EDT 2023
```

`dmesg` output:

```
[  +0.000000] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=systemd-oomd.service,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-80.scope,task=tantivy,pid=852338,uid=1000
[  +0.000015] Out of memory: Killed process 852338 (tantivy) total-vm:204738096kB, anon-rss:108496444kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:396124kB oom_score_adj:0
[  +4.669507] oom_reaper: reaped process 852338 (tantivy), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```


I don't think the problem is specific to the computer itself, especially considering the processing rate dropped down to 0.15Mb/s right before the process was killed and considering it's my daily workstation and I don't see spurious OOM killer problems.

I've attached what I little bit of information I can think of that might be useful.  I'll hold onto the output directory for a bit if someone would like it to examine it.  It's 436G uncompressed.  I'm also happy to share the input data.

Let me know if there's any other information I can provide or if it looks like I've done something wrong.  This isn't a mission critical thing for me.  I was really just curious to see how long tantivy would take to index 1.5T of really dense text, and then mess around with searching it a bit.

[full-output.log.gz](https://github.com/quickwit-oss/tantivy-cli/files/12464887/full-output.log.gz)
[dir-contents.txt.gz](https://github.com/quickwit-oss/tantivy-cli/files/12464900/dir-contents.txt.gz)

**Comment by eeeebbbbrrrr** [comment]
I used the main branch of `tantivy-cli`.  Is that not up-to-date with `tantivy:main`?

**Comment by PSeitz** [comment]
Only up to date with the latest release https://github.com/quickwit-oss/tantivy-cli/blob/main/Cargo.toml#L30

**Comment by eeeebbbbrrrr** [comment]
It finished after 7h 18m with the index consuming 2.6T.  I'll let you decide when to close this issue since this repo technically still has the bug.

Maybe when the next version of `tantivy` is released, a corresponding `tantivy-cli` can be published on crates.io too?

**Comment by PSeitz** [comment]
tantivy `0.21` has been released and `tantivy-cli` updated

**Comment by eeeebbbbrrrr** [comment]
Do you intend to publish a new tantivy-cli to crates.io?  It's still at 0.19.0

**Comment by PSeitz** [comment]
I released `tantivy-cli` 0.21
